components/Common/Pagination/useGetPageElements.tsx

```diff
@@ -0,0 +1,146 @@
+'use client';
+import Link from 'next/link';
+import { useIntl } from 'react-intl';
+
+import styles from './index.module.css';
+
+import type { PaginationProps } from '.';
+
+type ParsedPage = PaginationProps['pages'][number] & {
+  pageNumber: number;
+};
+
+function parsePages(pages: PaginationProps['pages']): ParsedPage[] {
+  return pages.map(({ url }, index) => ({
+    url,
+    pageNumber: index + 1,
+  }));
+}
+
+const ellipsis = (
+  <span aria-hidden="true" className={styles.ellipsis}>
+    ...
+  </span>
+);
+
+function createPageElements({
+  parsedPages,
+  currentPage,
+  intl,
+}: Pick<PaginationProps, 'currentPage'> & {
+  parsedPages: ParsedPage[];
+  intl: ReturnType<typeof useIntl>;
+}) {
+  return parsedPages.map(({ url, pageNumber }) => {
+    return (
+      <li
+        key={pageNumber}
+        aria-setsize={parsedPages.length}
+        aria-posinset={pageNumber}
+      >
+        <Link
+          prefetch={false}
+          href={url}
+          aria-label={intl.formatMessage(
+            {
+              id: 'components.common.pagination.pageLabel',
+            },
+            {
+              pageNumber,
+            }
+          )}
+          className={styles.listItem}
+          {...(pageNumber === currentPage && { 'aria-current': 'page' })}
+        >
+          <span>{pageNumber}</span>
+        </Link>
+      </li>
+    );
+  });
+}
+
+export const useGetPageElements = (
+  currentPage: PaginationProps['currentPage'],
+  pages: PaginationProps['pages'],
+  currentPageSiblingsCount: number
+) => {
+  const intl = useIntl();
+
+  const parsedPages = parsePages(pages);
+  const totalPages = parsedPages.length;
+  /**
+   * The combination of firstElement + currentElement + lastElement + 2 ellipses
+   */
+  const minimumElements = 5;
+
+  const minimumAmountOfPages = currentPageSiblingsCount + minimumElements;
+
+  if (totalPages <= minimumAmountOfPages) {
+    return createPageElements({ parsedPages, currentPage, intl });
+  }
+
+  const leftSiblingIndex = Math.max(
+    currentPage - currentPageSiblingsCount - 1,
+    1
+  );
+  const rightSiblingIndex = Math.min(
+    currentPage + currentPageSiblingsCount,
+    totalPages
+  );
+
+  const shouldShowLeftDots = leftSiblingIndex > 2;
+  const shouldShowRightDots = rightSiblingIndex < totalPages - 2;
+
+  const firstPageIndex = 0;
+  const lastPageIndex = totalPages - 1;
+
+  if (!shouldShowLeftDots && shouldShowRightDots) {
```

Comment:
I see a lot of duplicated logic / very similar logic. I wonder if recursivity could be used here or some sort of design pattern to simplify the logic and reduce the overall gigantic size of this Hook.

There are many moving parts here and little to none documentation. Like why "3 + 2 * something"... From where is that even coming from?

Reply:
Thank you. As mentioned in [this comment](https://github.com/nodejs/nodejs.org/pull/5971#discussion_r1349698931), the logic is inherently complex, and so the code, even with comments and after replacing the `3` and `2` magic numbers, will still be naturally "complex."

I thought about recursion but I thought it would not make sense here, as we don't need to traverse the elements of the array. We just need to make calculations based on the array's size, and other variables at play.

I will modify the code, remove magic numbers, add more comments to document what each step is doing, and try to reduce the amount of lines the custom hook has.

I agree with you in that it's complex, though, so will try my best to make it all simpler.

Reply:
> Thank you. As mentioned in this comment, the logic is inherently complex, and so the code, even with comments and after replacing the 3 and 2 magic numbers, will still be naturally "complex."

Right, understandable. Don't forget to make these magic numbers actual constants. So it's easier to update their usages, etc.